server/metrics/requests.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+
+	"github.com/uber-go/tally"
+	"google.golang.org/grpc"
+)
+
+const (
+	ServerRequestsReceivedTotal      = "server_requests_received_total"
+	ServerRequestsHandledTotal       = "server_requests_handled_total"
+	ServerRequestsErrorTotal         = "server_requests_error_total"
+	ServerRequestsOkTotal            = "server_requests_ok_total"
+	ServerRequestsHandlingTimeBucket = "server_requests_handling_time_bucket"
+)
+
+var (
+	ServerRequestsCounterNames = []string{
+		ServerRequestsReceivedTotal,
+		ServerRequestsHandledTotal,
+		ServerRequestsErrorTotal,
+		ServerRequestsOkTotal,
+	}
+	ServerRequestsHistogramNames = []string{
+		ServerRequestsHandlingTimeBucket,
+	}
+)
+
+type GrpcEndPointMetadata struct {
+	grpcServiceName string
+	grpcMethodName  string
+	grpcTypeName    string
+}
+
+func NewGrpcEndpointMetadata(grpcServiceName string, grpcMethodName string, grpcTypeName string) GrpcEndPointMetadata {
+	return GrpcEndPointMetadata{grpcServiceName: grpcServiceName, grpcMethodName: grpcMethodName, grpcTypeName: grpcTypeName}
+}
+
+func (g *GrpcEndPointMetadata) getTags() map[string]string {
+	return map[string]string{
+		"grpc_method":  g.grpcMethodName,
+		"grpc_service": g.grpcServiceName,
+		"grpc_type":    g.grpcTypeName,
+	}
+}
+
+func (g *GrpcEndPointMetadata) getFullMethod() string {
+	return fmt.Sprintf("/%s/%s", g.grpcServiceName, g.grpcMethodName)
+}
+
+func getGrpcEndPointMetadata(svcName string, methodInfo grpc.MethodInfo) GrpcEndPointMetadata {
+	var endpointMetadata GrpcEndPointMetadata
+	if methodInfo.IsServerStream {
+		endpointMetadata = NewGrpcEndpointMetadata(svcName, methodInfo.Name, "stream")
+	} else {
+		endpointMetadata = NewGrpcEndpointMetadata(svcName, methodInfo.Name, "unary")
+	}
+	return endpointMetadata
+
+}
+
+func InitServerRequestCounters(svcName string, methodInfo grpc.MethodInfo) {
+	endpointMetadata := getGrpcEndPointMetadata(svcName, methodInfo)
+	fullMethodName := endpointMetadata.getFullMethod()
+	tags := endpointMetadata.getTags()
+
+	for _, counterName := range ServerRequestsCounterNames {
+		counter := ServerRequestCounter{
+			Name:    counterName,
+			Tags:    tags,
+			Counter: Root.Tagged(tags).Counter(counterName),
+		}
+		_, ok := ServerRequestCounters[fullMethodName]
+		if !ok {
```

Comment:
This two lines can be combined:
```
if _, ok := ServerRequestCounters[fullMethodName]; !ok {
...
```

Reply:
Great catch!

Reply:
Actually if I combine these 2 line it panics, because that part of the map is not initialized yet.

Reply:
Scratch that, it's combined now.